### ISSUE TITLE
Test that all private suffixes end in ICANN suffixes

### ISSUE THREAD

**Issue by jsha** [publicsuffix/list#160]
I believe it is the case that all private suffixes end in an ICANN suffix. Is that accurate? If so, is it a guaranteed property of the PSL, and is it part of the automated tests?

Thanks,
Jacob

**Comment by sleevi** [comment]
> Is that accurate?

Pedantically, no. ICANN gTLDs are a subset of the IANA Root Zone Database (which includes ccTLDs and sponsored or technical domains, such as .mil or .arpa). However, it can't be said that the PSL will only contain domains in the IANA RZD, since we accept ICANN gTLDs once they have executed a contract with ICANN, which happens before delegation (aka when they appear in the IANA RZD) - between 90 to 180 days, but that can increase further.

The PSL also contains some deprecated ccTLDs that were removed from the IANA RZD but that we haven't gotten around to pruning, as well as deprecated gTLDs that ICANN removed (most recently, .doosan).

**Comment by sleevi** [comment]
Summoning @gerv and @weppos in the event they weren't monitoring this.

**Comment by weppos** [comment]
What @sleevi said is accurate. Another example are non-ICANN TLDs such as .IO or .UK that are in the IANA RZD and therefore we list them in the "ICANN" section. Technically speaking, the "ICANN" section should probably be called IANA section.

It's also worth to mention that the private section is not guaranteed to contain all the IANA root zone domains, mostly because we don't sync it automatically (for example we don't include yet the latest batch of gTLDs). In theory, it's not even required because of the rule set in the formal PSL algorithm that states "if no match, "*" is assumed".

Conversely, it's guaranteed that if a suffix is in the ICANN section, it is/was a IANA RZD TLD. As @sleevi pointed out, we still have some deprecated TLDs, but I expect them to be removed (in fact, unless any objection I'd go ahead and remove `.doosan`). For example, CentralNic suffixes are not in the ICANN section because they are private domains (although they tend to be advertised as TLDs).

@jsha if you elaborate what you have in mind, we may be able to help you to better determine if the PSL can solve the problem or not.

**Comment by gerv** [comment]
Yes, the ICANN section should probably be the IANA section. I was young and foolish ;-)

However, it _should_ be a property (or, at least, notable if not) that all the domains in the PRIVATE section should end in a TLD frm the ICANN section. Wouldn't you agree?

Don't remove .doosan, because that section of the PSL is automatically generated, and it'll just get added back next time we take an update. Let's wait for ICANN to remove it from newgtlds.csv.

**Comment by jsha** [comment]
> However, it should be a property (or, at least, notable if not) that all the domains in the PRIVATE section should end in a TLD frm the ICANN section. Wouldn't you agree?

Yes, this is the property I was thinking of.

Ryan, your correction regarding ICANN gTLDs vs IANA RZD is informative, thanks! I had a misunderstanding that ICANN managed all the TLDs in the ICANN section.

In Boulder we have a sanity check that [names end in a suffix from the ICANN section](https://github.com/letsencrypt/boulder/blob/master/policy/policy-authority.go#L182) before we even attempt to validate. I added a function to Go's `publicsuffix` package to specifically look for suffixes from the ICANN section (vs all public suffixes), but after [review feedback](https://go-review.googlesource.com/#/c/16712/) am more or less convinced that it's unnecessary. So I wanted to double-check that "names ending in a public suffix" are a subset of "names ending in a public suffix from the ICANN section."

**Comment by gerv** [comment]
So I think we could have a test that the property named above is true, if someone wants to write one.

**Comment by weppos** [comment]
@gerv there's a pending discussion about tests. I need your :+1: in #130 to merge a code that have a copyright belonging to @rockdaboot 

Once either that PR or #123 (which was the alternative I temporarily put on hold) is merged, we can work on the test as well.
